### PR TITLE
3D Graph SLAM +3D lidar odom / 2D Laser SLAM + 2D lidar odom

### PR DESCRIPTION
**Pull request by ugsfume**
The main goal of this build is to implement SLAM using 3D point cloud (for ADS-DV) and 2D laser (for small-scale car). Odometry sources are also provided by these sensors, to replace or compliment wheel-based odometry. 
General Software Pipeline: Perception -> Cone Recognition + Cluster Detection -> Cone Mapping -> SLAM (optional for simulation) -> Path Planning -> Path Follower

Major Features:
- 3D Graph SLAM
- 2D Laser SLAM
- 3D lidar odom
- 2D lidar odom
- Cone distance check function: instead of relying on limiting point cloud distance cut-off to ignore cones that are far (only works in simulation), a function computes distance of cones from robot and reject those out of the defined range (works in both simulation and real world)
- Implemented skidpad routine, provided by the path planning library

Minor Features:
- Tweaked /cmd_vel_to_ackermann_drive wheelbase configuration to fit ADS-DV and small-scale car
- ARM64 patch for hdl_graph_slam (Frank's solution)

TODO:
- small-scale car communication WIP
- lap counting
- mission specific behaviours
- ADS-DV API

Instructions:

Autonomous System Launch Sequence (full-scale ADS-DV model, with absolute odometry):
1. roscore
2. roslaunch ackermann_vehicle_description ads_dv_sim.launch
3. roslaunch ultralytics_ros tracker_with_cloud.launch
4. roslaunch ultralytics_ros cone_mapper.launch
5. roslaunch ackermann_vehicle_navigation track_follower.launch

Do the following to disable GPU Acceleration:
1. Go to gra/.devcontainer/docker-compose.yml and comment the deploy section, then rebuild the Docker image in VScode
2. Go to gra/ros/cone_detection/launch/tracker_with_cloud.launch
	Comment line 17 and uncomment line 15
3. This step reduces Gazebo CPU usage by scaling the simulation time. The simulator will appear slower but CPU load will be reduced significantly.
	Go to gra/ros/ackermann_vehicle_gazebo/worlds/gazebo_world_building/track_small_features.world
	Line 76: <real_time_update_rate>700</real_time_update_rate>
	Reduce the value to reduce CPU load. 1000 is the default value where the simulation time unit = real time unit. When the value is 500, 1 simulation second = 2 real time second.
	Start with 700 and reduce gradually if the cone recognition process is unable to stablise. Unstable process can be observed by the cones swerving in Rviz when the car is steering.

Note to self: add these instructions to README or wiki

2d_slam_grid
3d_graph_slam

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -175,9 +158,28 @@ WORKDIR /opt/fsd_path_planning
 RUN git clone https://github.com/ugsfume/ft-fsd-path-planning.git .
 RUN python3 -m pip install .
 
+# # VSLAM dependencies and software
```
Comment (daniel-bara) [inline .devcontainer/Dockerfile:161]: Why do we need these commented-out lines?

File: ros/ackermann_vehicle_description/launch/rviz.launch
Comment (daniel-bara) [inline ros/ackermann_vehicle_description/launch/rviz.launch:1]: Empty file

File: ros/ackermann_vehicle_description/scripts/sensors.rviz
Comment (daniel-bara) [inline ros/ackermann_vehicle_description/scripts/sensors.rviz:1]: Empty file

File: ros/ackermann_vehicle_description/scripts/time_jump.py
Comment (daniel-bara) [inline ros/ackermann_vehicle_description/scripts/time_jump.py:1]: Empty file

File: ros/ackermann_vehicle_description/scripts/sensors_3rdP.rviz
Comment (daniel-bara) [inline ros/ackermann_vehicle_description/scripts/sensors_3rdP.rviz:1]: What's `3rdP`?

File: ros/ackermann_vehicle_navigation/launch/path_follower.launch
Comment (daniel-bara) [inline ros/ackermann_vehicle_navigation/launch/path_follower.launch:1]: Empty files everywhere

File: ros/ackermann_vehicle_navigation/launch/track_follower.launch [hunk 2]
```diff
@@ -1,9 +1,11 @@
 <launch>
   
   <!-- This node computes the path for the car to follow based on position of cones and car position + direction -->
-  <node name="track_pathfinder" pkg="ackermann_vehicle_navigation" type="track_pathfinder.py" output="screen"/>
+  <node name="track_pathfinder" pkg="ackermann_vehicle_navigation" type="track_pathfinder.py" output="screen" launch-prefix="nice -n -20"/>
 
   <!-- This node compute the transformation required for the car to achieve poses in /path. Output is /cmd_vel -->
-  <node name="path_follower" pkg="ackermann_vehicle_navigation" type="path_follower.py" output="screen"/>
+  <node name="path_follower" pkg="ackermann_vehicle_navigation" type="path_follower.py" output="screen" launch-prefix="nice -n -19"/>
```
Comment (daniel-bara) [inline ros/ackermann_vehicle_navigation/launch/track_follower.launch:7]: So we want the path follower at a lower priority than the path finder?

File: ros/ackermann_vehicle_navigation/scripts/angle_converter.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python3
+
+import rospy
+from geometry_msgs.msg import Twist
+from std_msgs.msg import UInt16
+from ackermann_msgs.msg import AckermannDrive
+
+def convert_to_angle(steering_angle, speed):
+    motor_angle=int(99+80*speed)
```
Comment (daniel-bara) [inline ros/ackermann_vehicle_navigation/scripts/angle_converter.py:9]: These would be more readable if we defined a `constrain(n)` and an `interpolate(x, a, b, c, d)` function (or imported them from somewhere).

File: ros/ackermann_vehicle_navigation/scripts/track_pathfinder.py [hunk 4]
```diff
@@ -51,7 +51,7 @@ def publish_path(self, path, header):
         ros_path.header = Header(stamp=rospy.Time.now(), frame_id='world')
 
         # Start from the 4th waypoint, skipping the first three
-        for point in path[3:]:
+        for point in path[4:]:
```
Comment (daniel-bara) [inline ros/ackermann_vehicle_navigation/scripts/track_pathfinder.py:54]: Comment is outdated now

File: .devcontainer/Dockerfile [hunk 5]
```diff
@@ -175,9 +158,28 @@ WORKDIR /opt/fsd_path_planning
...
+# RUN cmake ..
+# RUN sudo make -j4 install
+
+# # google-glog + gflags
+# RUN /usr/local/bin/apt-get-wrapper.sh libgoogle-glog-dev libgflags-dev
+# # Use ATLAS for BLAS & LAPACK
+# RUN /usr/local/bin/apt-get-wrapper.sh libatlas-base-dev
+# # SuiteSparse (optional)
+# RUN /usr/local/bin/apt-get-wrapper.sh libsuitesparse-dev
+
```
Comment (daniel-bara) [inline .devcontainer/Dockerfile:176]: Why do we need these commented-out lines?

File: ros/ackermann_vehicle_navigation/launch/path_follower.launch
Comment (ugsfume) [reply ros/ackermann_vehicle_navigation/launch/path_follower.launch:1]: The files are not empty, I'm not sure why github is showing that. You could see them with "view file"

**Comment by daniel-bara** [comment]
> Note to self: add these instructions to README or wiki

File: ros/ackermann_vehicle_navigation/launch/path_follower.launch
Comment (daniel-bara) [reply ros/ackermann_vehicle_navigation/launch/path_follower.launch:1]: If you change the (I think it's called header) with chown, it produces a diff in git

File: ros/ackermann_vehicle_navigation/launch/track_follower.launch (hunk 2 above)
Comment (daniel-bara) [reply ros/ackermann_vehicle_navigation/launch/track_follower.launch:7]: And we're using -20 and -19 here, right?

File: ros/ackermann_vehicle_description/scripts/sensors_3rdP.rviz
Comment (daniel-bara) [reply ros/ackermann_vehicle_description/scripts/sensors_3rdP.rviz:1]: waiting for the removal

Comment (ugsfume) [reply ros/ackermann_vehicle_description/scripts/sensors_3rdP.rviz:1]: It was removed

File: ros/ackermann_vehicle_navigation/launch/track_follower.launch (hunk 2 above)
Comment (ugsfume) [reply ros/ackermann_vehicle_navigation/launch/track_follower.launch:7]: Yes

**Review by daniel-bara — APPROVED** [review]

**Pull request merged**